src/source.rs

```diff
@@ -0,0 +1,146 @@
+//! Source trait and implementations for loading configuration data
+
+use crate::Result;
+use std::env;
+
+/// Trait for configuration sources that can provide values for specific fields
+/// 
+/// Sources should be efficient and handle arbitrarily large datasets by providing
+/// values only for requested fields rather than loading entire datasets.
+pub trait Source {
+    /// Get a value for a specific field
+    /// 
+    /// Returns None if the field is not available in this source.
+    /// The field name should be a simple identifier without complex paths.
+    fn get_field(&self, field: &str) -> Result<Option<serde_json::Value>>;
+}
+
+/// Source implementation for environment variables
+pub struct EnvSource {
+    prefix: Option<String>,
+}
+
+impl EnvSource {
+    /// Create a new EnvSource that reads all environment variables
+    pub fn new() -> Self {
+        Self { prefix: None }
+    }
+
+    /// Create a new EnvSource with a prefix filter
+    pub fn with_prefix<S: Into<String>>(prefix: S) -> Self {
+        Self {
+            prefix: Some(prefix.into()),
+        }
+    }
+
+    /// Convert a field name to an environment variable key
+    fn field_to_env_key(&self, field: &str) -> String {
+        if let Some(ref prefix) = self.prefix {
+            format!("{}_{}", prefix, field)
+        } else {
+            field.to_string()
+        }
+    }
+}
+
+impl Default for EnvSource {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Source for EnvSource {
+    fn get_field(&self, field: &str) -> Result<Option<serde_json::Value>> {
```

Comment:
Env naturally should allow Config's field to be parsed using field's FromStr implementation if available.
And only if it isn't, or configured otherwise - parsed from JSON using `serde_json::from_str`, not with roundtrip through `Value`.